### PR TITLE
Remove the ObjectWrapper class

### PR DESCRIPTION
**Pull request by rolfbjarne**
The ObjectWrapper class uses a dirty hack to convert between a native `MonoObject*` pointer and a managed object: it uses a struct with two fields, one `IntPtr` and the other `object`, that have the same field offset. This means that you could write the `MonoObject*` value to the `IntPtr` field, and read the managed object back from the `object` field. This is not a very portable hack, and in fact not allowed in .NET Core. Going forward, the only supported way to convert between managed objects and their corresponding native `MonoObject*` objects is to use `GCHandle`s, which have an API that allows us to both create and read the `GCHandle` from either side, getting either a managed object or a native `MonoObject*`.

This required extensive changes in the managed methods/delegates that native code calls, to pass `GCHandle`s around instead of `MonoObject*` pointers.

This PR is best reviewed in commit-sized chunks.

### CODE REVIEW

File: src/ObjCRuntime/Runtime.cs [hunk 1]
```diff
@@ -707,24 +729,30 @@ static IntPtr GetINativeObject_Static (IntPtr ptr, bool owns, uint iface_token,
...
+			return GCHandle.ToIntPtr (GCHandle.Alloc (GetNSObject (ptr, type, MissingCtorResolution.ThrowConstructor1NotFound, true, out created)));
 		}
 
-		static void Dispose (IntPtr mobj)
+		static void Dispose (IntPtr gchandle)
 		{
-			((IDisposable) ObjectWrapper.Convert (mobj)).Dispose ();
+			var handle = GCHandle.FromIntPtr (gchandle);
+			((IDisposable) handle.Target).Dispose ();
+			handle.Free ();
```
Comment (spouliot) [inline src/ObjCRuntime/Runtime.cs:748]: `Dispose` could throw and (if handled by callers) the new code would leak the `GCHandle`

File: src/ObjCRuntime/Runtime.cs [hunk 2]
```diff
@@ -378,9 +378,11 @@ static IntPtr ConvertSmartEnumToNSString (IntPtr value_handle)
 
 		// value: native NSString *
 		// returns: GCHandle to a (smart) enum value. Caller must free the GCHandle.
-		static IntPtr ConvertNSStringToSmartEnum (IntPtr value, IntPtr type)
+		static IntPtr ConvertNSStringToSmartEnum (IntPtr value, IntPtr type_handle)
 		{
-			var smart_type = (Type) ObjectWrapper.Convert (type);
+			var handle = GCHandle.FromIntPtr (type_handle);
+			var smart_type = (Type) handle.Target;
+			handle.Free ();
```
[the remark this answers is not in the document]
Comment (rolfbjarne) [reply src/ObjCRuntime/Runtime.cs:385]: Yes, there's a bunch of TODOs here: https://github.com/xamarin/xamarin-macios/pull/8463/commits/19ca05f1e945a8214bea70580d86bc049120cddd

[the remark this answers is not in the document]
Comment (rolfbjarne) [reply src/ObjCRuntime/Runtime.cs]: This has been reworked now.

File: src/ObjCRuntime/Runtime.cs (hunk 1 above)
Comment (rolfbjarne) [reply src/ObjCRuntime/Runtime.cs]: This has been reworked now.

**Pull request merged**